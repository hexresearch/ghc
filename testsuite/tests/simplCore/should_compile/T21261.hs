module T21261 where

-- README: The convention here is that bindings starting with 'yes' should be
-- eta-reduced and become trivial, while bindings starting with 'no' should not
-- be eta-reduced.

f1 :: (Int -> Int -> Int) -> Int
f1 c = c 1 2 + c 3 4
{-# NOINLINE f1 #-}
yes1 :: (Int -> Int -> Int) -> Int
yes1 c = f1 (\x -> c x)

f2 :: (Int -> Int -> Int) -> Int
f2 c = c 1 `seq` c 3 4
{-# NOINLINE f2 #-}
yes1or2 :: (Int -> Int -> Int) -> Int
yes1or2 c = f2 c

f3 :: (Int -> Int -> Int) -> Int
f3 c = c 1 2 + c 3 4
{-# NOINLINE f3 #-}
yes2 :: (Int -> Int -> Int) -> Int
yes2 c = f3 (\x y -> c x y)

f4 :: (Int -> Int -> Int -> Int) -> Int
f4 c = c 1 2 `seq` c 3 4 `seq` 42
{-# NOINLINE f4 #-}
no3 :: (Int -> Int -> Int -> Int) -> Int
no3 c = f4 (\x y z -> c x y z)

f5 :: (Int -> Int -> Int) -> Maybe Int
f5 c = Just (c 1 2 + c 3 4)
{-# NOINLINE f5 #-}
yes2_lazy :: (Int -> Int -> Int) -> Maybe Int
yes2_lazy c = f5 (\x y -> c x y)

f6 :: (Int -> Int -> Int) -> Maybe Int
f6 c = Just (c 1 `seq` c 3 4)
{-# NOINLINE f6 #-}
no2_lazy :: (Int -> Int -> Int) -> Maybe Int
no2_lazy c = f6 (\x y -> c x y)

f7 :: (Int -> Int -> Int) -> Int
f7 c = c 1 `seq` c 2 3
{-# NOINLINE f7 #-}
not_quite_eta :: (Int -> Int -> Int) -> Int
not_quite_eta c = f7 (\x y -> c x y)
