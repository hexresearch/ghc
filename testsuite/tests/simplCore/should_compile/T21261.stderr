
==================== Tidy Core ====================
Result size of Tidy Core
  = {terms: 192, types: 201, coercions: 0, joins: 0/0}

lvl = I# 3#

lvl1 = I# 4#

lvl2 = I# 1#

f2 = \ c -> case c lvl2 of { __DEFAULT -> c lvl lvl1 }

yes1or2 = f2

lvl3 = I# 2#

$wf4
  = \ c ->
      case c lvl2 lvl3 of { __DEFAULT ->
      case c lvl lvl1 of { __DEFAULT -> 42# }
      }

f4 = \ c -> case $wf4 c of ww { __DEFAULT -> I# ww }

no3
  = \ c ->
      case $wf4 (\ x y z -> c x y z) of ww { __DEFAULT -> I# ww }

$wf6 = \ c -> (# case c lvl2 of { __DEFAULT -> c lvl lvl1 } #)

f6 = \ c -> case $wf6 c of { (# ww #) -> Just ww }

no2_lazy
  = \ c -> case $wf6 (\ x y -> c x y) of { (# ww #) -> Just ww }

f7 = \ c -> case c lvl2 of { __DEFAULT -> c lvl3 lvl }

not_quite_eta = \ c -> f7 (\ x y -> c x y)

$wf5
  = \ c ->
      (# case c lvl2 lvl3 of { I# x ->
         case c lvl lvl1 of { I# y -> I# (+# x y) }
         } #)

f5 = \ c -> case $wf5 c of { (# ww #) -> Just ww }

yes2_lazy
  = \ c -> case $wf5 (\ x y -> c x y) of { (# ww #) -> Just ww }

$wf3
  = \ c ->
      case c lvl2 lvl3 of { I# x ->
      case c lvl lvl1 of { I# y -> +# x y }
      }

f3 = \ c -> case $wf3 c of ww { __DEFAULT -> I# ww }

yes2 = f3

$wf1
  = \ c ->
      case c lvl2 lvl3 of { I# x ->
      case c lvl lvl1 of { I# y -> +# x y }
      }

f1 = \ c -> case $wf1 c of ww { __DEFAULT -> I# ww }

yes1 = f1



